File: 2.addTwoNumbers/main.cpp
#include <iostream>
using namespace std;

struct ListNode {
     int val;
     ListNode *next;

     ListNode() : val(0), next(nullptr) {}

     ListNode(int x) : val(x), next(nullptr) {}

     ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *ans = new ListNode;
        ListNode *tail = ans;
        int c = 0;
        while (l1 != nullptr && l2 != nullptr) {
            int temp = l1->val + l2->val + c;
            if (temp >= 10) {
                temp -= 10;
                c = 1;
            } else {
                c = 0;
            }
            ListNode* next_node = new ListNode(temp);
            tail->next = next_node;
            tail = tail->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1 != nullptr) {
            int temp = l1->val + c;
            if (temp >= 10) {
                temp -= 10;
                c = 1;
            } else {
                c = 0;
            }
            ListNode* next_node = new ListNode(temp);
            tail->next = next_node;
            tail = tail->next;
            l1 = l1->next;
        }
        while (l2 != nullptr) {
            int temp = l2->val + c;
            if (temp >= 10) {
                temp -= 10;
                c = 1;
            } else {
                c = 0;
            }
            ListNode *next_node = new ListNode(temp);
            tail->next = next_node;
            tail = tail->next;
            l2 = l2->next;
        }
        if (c==1){
            ListNode *next_node = new ListNode(c);
            tail->next = next_node;
            tail = tail->next;
        }
        ans = ans->next;
        return ans;
    }
};


int main() {
    ListNode l17(9, nullptr);
    ListNode l16(9, &l17);
    ListNode l15(9, &l16);
    ListNode l14(9, &l15);
    ListNode l13(9, &l14);
    ListNode l12(9,&l13);
    ListNode l1(9,&l12);
    ListNode l24(9, nullptr);
    ListNode l23(9, &l24);
    ListNode l22(9,&l23);
    ListNode l2(9,&l22);
    Solution solution;
    ListNode* ans = solution.addTwoNumbers(&l1,&l2);
    return 0;
}

File: 2.addTwoNumbers/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// 快手二面
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var ret *ListNode
	var tail *ListNode
	c := 0
	for l1 != nil || l2 != nil {
		num1 := 0
		if l1 != nil {
			num1 = l1.Val
			l1 = l1.Next
		}
		num2 := 0
		if l2 != nil {
			num2 = l2.Val
			l2 = l2.Next
		}
		temp := num1 + num2 + c
		c = temp / 10
		if ret == nil {
			ret = &ListNode{Val: temp % 10}
			tail = ret
		} else {
			tail.Next = &ListNode{Val: temp % 10}
			tail = tail.Next
		}
	}
	if c != 0 {
		tail.Next = &ListNode{Val: c}
	}
	return ret
}

func main() {
	head1 := &ListNode{
		Val: 2,
		Next: &ListNode{
			Val:  4,
			Next: &ListNode{Val: 3},
		},
	}
	head2 := &ListNode{
		Val: 5,
		Next: &ListNode{
			Val:  6,
			Next: &ListNode{Val: 4},
		},
	}
	ret := addTwoNumbers(head1, head2)
	fmt.Printf("%+v\n", ret)
}
